File: NET_4.5.2/C1.WPF.FlexGrid/VB/CollectionViewFilter/CollectionViewFilter/ViewFilter.vb
Imports System.Text

Namespace CollectionViewFilter
	Public Class ViewFilter
		' ** fields
		Private _view As System.ComponentModel.ICollectionView
		Private _filterExpression As String
		Private _dt As DataTable

		' ** ctor
		Public Sub New(ByVal view As System.ComponentModel.ICollectionView)
			_view = view
		End Sub

		' ** object model
		Public Property FilterExpression() As String
			Get
				Return _filterExpression
			End Get
			Set(ByVal value As String)
				_filterExpression = value
				UpdateFilter()
				_view.Filter = Nothing
				If Not String.IsNullOrEmpty(_filterExpression) Then
					_view.Filter = AddressOf FilterPredicate
				End If
			End Set
		End Property

		' ** implementation
		Private Function FilterPredicate(ByVal obj As Object) As Boolean
			' populate the row
			Dim row = _dt.Rows(0)
			For Each pi In obj.GetType().GetProperties()
				row(pi.Name) = pi.GetValue(obj, Nothing)
			Next pi

			' compute the expression
			Return CBool(row("_filter"))
		End Function
		Private Sub UpdateFilter()
			_dt = Nothing
			If _view.CurrentItem IsNot Nothing AndAlso (Not String.IsNullOrEmpty(_filterExpression)) Then
				' build/rebuild data table
				Dim dt = New DataTable()
				For Each pi In _view.CurrentItem.GetType().GetProperties()
					dt.Columns.Add(pi.Name, pi.PropertyType)
				Next pi

				' add calculated column
				dt.Columns.Add("_filter", GetType(Boolean), _filterExpression)

				' create a single row for evaluating expressions
				If dt.Rows.Count = 0 Then
					dt.Rows.Add(dt.NewRow())
				End If

				' done, save table
				_dt = dt
			End If
		End Sub
	End Class
End Namespace

File: NET_4.6.2/C1.WPF.FlexGrid/CS/CollectionViewFilter/CollectionViewFilter/ViewFilter.cs
using System;
using System.Data;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace CollectionViewFilter
{
    public class ViewFilter
    {
        // ** fields
        System.ComponentModel.ICollectionView _view;
        string _filterExpression;
        DataTable _dt;

        // ** ctor
        public ViewFilter(System.ComponentModel.ICollectionView view)
        {
            _view = view;
        }

        // ** object model
        public string FilterExpression
        {
            get { return _filterExpression; }
            set
            {
                _filterExpression = value;
                UpdateFilter();
                _view.Filter = null;
                if (!string.IsNullOrEmpty(_filterExpression))
                {
                    _view.Filter = FilterPredicate;
                }
            }
        }

        // ** implementation
        bool FilterPredicate(object obj)
        {
            if (_dt == null)
                return false;
            // populate the row
            var row = _dt.Rows[0];
            foreach (var pi in obj.GetType().GetProperties())
            {
                row[pi.Name] = pi.GetValue(obj, null);
            }

            // compute the expression
            return (bool)row["_filter"];
        }

        void UpdateFilter()
        {
            if (_dt == null)
            {
                // build data table
                if (_view.CurrentItem != null && !string.IsNullOrEmpty(_filterExpression))
                {
                    var dt = new DataTable();
                    foreach (var pi in _view.CurrentItem.GetType().GetProperties())
                    {
                        dt.Columns.Add(pi.Name, pi.PropertyType);
                    }

                    // add calculated column
                    dt.Columns.Add("_filter", typeof(bool), _filterExpression);

                    // create a single row for evaluating expressions
                    if (dt.Rows.Count == 0)
                    {
                        dt.Rows.Add(dt.NewRow());
                    }

                    // done, save table
                    _dt = dt;
                }
            }
            if ( _dt != null)
            {
                _dt.Columns["_filter"].Expression = _filterExpression;
            }
        }
    }
}
